collector/timex.go

```diff
@@ -0,0 +1,180 @@
+			nil, nil,
+		), prometheus.GaugeValue},
+		stabil: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "stabil"),
+			"Pulse per second stability.",
+			nil, nil,
+		), prometheus.CounterValue},
+		jitcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "jitcnt"),
+			"Pulse per second count of jitter limit exceeded events.",
+			nil, nil,
+		), prometheus.CounterValue},
+		calcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "calcnt"),
+			"Pulse per second count of calibration intervals.",
+			nil, nil,
+		), prometheus.CounterValue},
+		errcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "errcnt"),
+			"Pulse per second count of calibration errors.",
+			nil, nil,
+		), prometheus.CounterValue},
+		stbcnt: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "stbcnt"),
+			"Pulse per second count of stability limit exceeded events.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		tai: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "tai"),
+			"International Atomic Time (TAI) offset.",
+			nil, nil,
+		), prometheus.GaugeValue},
+		syncStatus: typedDesc{prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "sync_status"),
+			"Is clock synchronized to a reliable server (1 = yes, 0 = no).",
+			nil, nil,
+		), prometheus.GaugeValue},
+	}, nil
+}
+
+func (c *timexCollector) Update(ch chan<- prometheus.Metric) error {
+	var syncStatus float64
+	var timex = new(syscall.Timex)
+
+	status, err := syscall.Adjtimex(timex)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve adjtimex stats: %v", err)
+	}
+
+	if status == timeError {
+		syncStatus = 0
+	} else {
+		syncStatus = 1
+	}
+	ch <- c.syncStatus.mustNewConstMetric(syncStatus)
+	ch <- c.offset.mustNewConstMetric(float64(timex.Offset) / 1000000)
+	ch <- c.freq.mustNewConstMetric(float64(timex.Freq))
+	ch <- c.maxerror.mustNewConstMetric(float64(timex.Maxerror) / 1000000)
+	ch <- c.esterror.mustNewConstMetric(float64(timex.Esterror) / 1000000)
+	ch <- c.status.mustNewConstMetric(float64(timex.Status))
+	ch <- c.constant.mustNewConstMetric(float64(timex.Constant))
+	ch <- c.precision.mustNewConstMetric(float64(timex.Precision) / 1000000)
+	ch <- c.tolerance.mustNewConstMetric(float64(timex.Tolerance))
+	ch <- c.tick.mustNewConstMetric(float64(timex.Tick) / 1000000)
+	ch <- c.ppsfreq.mustNewConstMetric(float64(timex.Ppsfreq))
+	ch <- c.jitter.mustNewConstMetric(float64(timex.Jitter) / 1000000)
+	ch <- c.shift.mustNewConstMetric(float64(timex.Shift))
+	ch <- c.stabil.mustNewConstMetric(float64(timex.Stabil))
+	ch <- c.jitcnt.mustNewConstMetric(float64(timex.Jitcnt))
+	ch <- c.calcnt.mustNewConstMetric(float64(timex.Calcnt))
+	ch <- c.errcnt.mustNewConstMetric(float64(timex.Errcnt))
+	ch <- c.stbcnt.mustNewConstMetric(float64(timex.Stbcnt))
+	ch <- c.tai.mustNewConstMetric(float64(timex.Tai))
```

Comment:
> +1 for more metrics and filter them later. We're just talking about a handful, right?

I'm not time-keeping expert at all (e.g. I don't understand `constant` value meaning), but are you sure you're not "overreacting" while exporting values that are not defined [without CONFIG_NTP_PPS](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/kernel/time/ntp.c?h=v4.12.10#n223)? I mean ppsfreq, jitter, shift, stabil, jitcnt, calcnt, errcnt, stbcnt.

I'm also unsure if `precision` and `tolerance` make sense, they're [hardcoded in Linux](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/kernel/time/ntp.c?h=v4.12.10#n757) and Linux is the only one that names `timex`-related function `adjtimex()`, so it's basically Linux-only collector right now.

Also, I'm unsure if [TAI offset](https://cr.yp.to/proto/utctai.html) is practical for node_exporter purposes as it basically counts number of leap seconds happened since epoch and I'm unsure if any userland sets it to non-zero value.

`status` is a bit field. Is it good to export it as `float64`?

Reply:
I'm even less of a time-keeping expert :)

What is CONFIG_NTP_PPS? If those values are known to be always 0, yes it doesn't make sense to expose them. If they are usually 0, exposing them is fine: Timeseries with constant values are pretty much for free in Prometheus.

Precision and tolerance probably make sense, since they could change between kernel versions and again they are pretty much free until they do.

Similar to the TAI offset. I'd say to keep them all but don't feel too strongly. So if we know for sure some of them are useless, we can remove them.

Reply:
> What is CONFIG_NTP_PPS?

That's [Linux kernel](https://github.com/torvalds/linux/blob/master/drivers/pps/Kconfig#L22) compile-time [option](http://cateee.net/lkddb/web-lkddb/NTP_PPS.html). What's important for node_exporter, that the option is usually turned off: either explicitly or as a side-effect of incompatibility with `CONFIG_NO_HZ_COMMON=y`. I've checked some of my machines running Ubuntu {14.04,16.04} and Debian jessie, none of them have CONFIG_NTP_PPS enabled. So I assume these variables are 0 for everyone besides those timekeepers who are running a Linux-based timeserver with custom kernel build.

> make sense, since they could change between kernel versions

Well. `statfs()` does not export `f_namelen` and `f_frsize`, and only `ST_RDONLY` is taken from `f_flags` for some reason. `kernel.random.entropy_avail` is exported, but `pool_size` is not, CPU microcode version is not exported as well :) I mean, I don't know where the line between "expose everything" and "noone needs that constant value" is, but current state is obviously not "expose everything".

I tried to use rule "expose the metrics if you know how to interpret that". I asked about `maxerror` as it kinda timestamps the event when time keeping daemon have set this field for the last time (as there is linear decay afterwards) and this value may be useful during incident response. I'm unsure if it really **is** useful as I have no experience managing time keepers, that's why I asked @kerolasa as I assumed that he had alike experience :)

Reply:
I dropped precision and tolerance. They do not seem to be worth collecting.

TAI offset should be updated by ntpd, but I don't see systemd-timesyncd setting it. This can be useful to someone who is trying to do science, and has trouble understanding why systems differ.

Status is in golang world int32, https://golang.org/pkg/syscall/#Timex I'm pretty sure float64 can handle that without issue. Notice that the statNano bit field operation is using untampered int32, so I am almost sure that works.

To question shoudl we have all the stuff clock server people might need. Originally I thought many of these are pointless to large majority. But then I was told metrics are cheap, so I added everything. Now couple most pointless are dropped, but there are still many that do not help in common time debugging situations.

@mattbostock told me that there has been some discussion to add example monitoring configuration somewhere in this project. In this case it would really help to tell to users unless they are running clock servers all pps_* values can be safely ignored. I also stick to my original opinion that for most setups collecting offset_seconds, frequency_adjustment, and sync_status is enough. Will there ever be case when knowing estimated_error_seconds will solve a mystery? Difficult to say, probably not, but by measuring for few years on thousands of servers conclusions can be made what is helpful. I assume dropping metrics in future is as fine as starting with too many.

Reply:
> TAI offset **should** be updated by ntpd

I asked as ntp-4.2.6.p5+dfsg-7+deb8u2 does not update it. Seems, it either needs non-default configuration or something else. :-(

> Will there ever be case when knowing estimated_error_seconds will solve a mystery?

I was asking about maxerror, not about esterror, as maxerror [decays](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/kernel/time/ntp.c?h=v4.12.10#n455). But I got your point, thanks.

Reply:
> Now couple most pointless are dropped, but there are still many that do not help in common time debugging situations.

I think that's a good start!

Reply:
@darkk thank you again. I hope the latest got all nano/micro second precisions right.

Reply:
Is it crunch time? Last call for review before merge, first proper try.